.github/workflows/deploy.yml

```diff
@@ -0,0 +1,155 @@
+name: Deploy
+on:
+  push:
+    branches: [main, release]
+    paths-ignore: ['**.md']
+  pull_request:
+    branches: [main]
+    paths-ignore: ['**.md']
+
+defaults:
+  run:
+    shell: bash
+
+permissions:
+  pull-requests: write
+
+jobs:
+  deploy:
+    name: Deploy
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v5
+      - uses: ./.github/workflows/setup
+      - name: Install wrangler globally
+        run: |
+          # the wrangler action expects a global installation, so we better reuse what we've in our project
+          wrangler_version=$(cat package.json | jq -r '.pnpm.overrides.wrangler')
+          pnpm install -g "wrangler@${wrangler_version}"
+
+      - uses: dorny/paths-filter@v3
+        id: changes
+        with:
+          filters: |
+            api: &api
+              - 'api/**'
+              - 'shared/**'
+            cdn: &cdn
+              - *api
+              - 'components/**'
+              - 'embed/**'
+              - 'cdn/**'
+            frontend:
+              - *api
+              - *cdn
+              - 'frontend/**'
+
+      - name: Get deploy mode
+        id: deploy-mode
+        run: |
+          mode="preview"
+          if [[ "$EVENT_NAME" == 'push' ]]; then
+            [[ "$GH_REF" == 'refs/heads/main' ]] && mode="staging"
+            [[ "$GH_REF" == 'refs/heads/release' ]] && mode="production"
+          fi
+          echo "mode=${mode}" >> $GITHUB_OUTPUT
+        env:
+          GH_EVENT_NAME: ${{ github.event_name }}
+          GH_REF: ${{ github.ref }}
```

Comment:
IIRC, inline interpolation can be used for `ref` and `event_name` in the bash script. If this is going to be something consistent - assigning to environment variables, I see no issue here.

Reply:
I wanted to have them available as variables, so used env variables - then I can copy the script out and test it locally.
Perhaps, instead of env vars, I'll see if I can instead use interpolation at the top level of the script - consistency is key.